File: 2024/JS/Day6/solutionA.js
const fs = require("fs");

const getLocation = (grid, char) => {
  let row = 0;
  let col = 0;

  for (let i = 0; i < grid.length; i++) {
    if (grid[i].includes(char)) {
      row = i;
      col = grid[i].indexOf(char);
      break;
    }
  }

  return { row, col };
};

const walk = (grid, startLoc) => {
  let newGrid = grid.map((row) => [...row]);
  const obstacle = "#";
  const directions = {
    N: { r: -1, c: 0 },
    E: { r: 0, c: 1 },
    S: { r: 1, c: 0 },
    W: { r: 0, c: -1 },
  };
  let rotations = {
    N: "E",
    E: "S",
    S: "W",
    W: "N",
  };
  let currentDirection = "N";
  let currentLocation = startLoc;

  let continueWalk = true;
  while (continueWalk) {
    let nextLocation = {
      row: currentLocation.row + directions[currentDirection].r,
      col: currentLocation.col + directions[currentDirection].c,
    };

    // break if next location is out of bounds
    if (
      nextLocation.row < 0 ||
      nextLocation.row >= grid.length ||
      nextLocation.col < 0 ||
      nextLocation.col >= grid[0].length
    ) {
      continueWalk = false;
      break;
    }

    // turn 90 degrees to the right if next location is an obstacle
    if (newGrid[nextLocation.row][nextLocation.col] === obstacle) {
      currentDirection = rotations[currentDirection];
      continue;
    }

    // mark the current location as visited
    newGrid[currentLocation.row][currentLocation.col] = "X";

    // move to next location
    currentLocation = nextLocation;
  }

  return newGrid.flatMap((row) => row).filter((cell) => cell === "X").length;
};

fs.readFile("../../data/day6.txt", (err, data) => {
  let input = data.toString();
  let grid = input.split("\n").map((row) => row.split(""));
  const { row, col } = getLocation(grid, "^");
  console.log(walk(grid, { row, col }));
});

File: 2024/Python/Day6/solutionA.py

def get_location(grid, char):
    for i in range(len(grid)):
        for j in range(len(grid[i])):
            if grid[i][j] == char:
                return {"row": i, "col": j}
    return None

def check_out_of_bounds(grid, location):
    if location["row"] < 0 or location["row"] >= len(grid) or location["col"] < 0 or location["col"] >= len(grid[0]):
        return True
    return False

def count_char(grid, char):
    count = 0
    for i in range(len(grid)):
        for j in range(len(grid[i])):
            if grid[i][j] == char:
                count += 1
    return count

def simulate_walk(grid):
    obstacle = "#"
    current_direction = "N"
    current_location = get_location(grid, "^")
    rotations = {
        "N": "E",
        "E": "S",
        "S": "W",
        "W": "N"
    }
    directions = {
        "N": {"r": -1, "c": 0},
        "E": {"r": 0, "c": 1},
        "S": {"r": 1, "c": 0},
        "W": {"r": 0, "c": -1}
    }

    continue_walk = True
    while continue_walk:
        grid[current_location["row"]][current_location["col"]] = "X"

        next_location = {
            "row": current_location["row"] + directions[current_direction]["r"],
            "col": current_location["col"] + directions[current_direction]["c"]
        }

        if check_out_of_bounds(grid, next_location):
            continue_walk = False
            break

        if grid[next_location["row"]][next_location["col"]] == obstacle:
            current_direction = rotations[current_direction]
            continue 

        current_location = next_location

    return count_char(grid, "X")

grid = []
with open("../../data/day6.txt") as file:
    for line in file:
        grid.append(list(line))

result = simulate_walk(grid)
print(result)
